### PR TITLE
using the wrong renderer's act() should warn

### PR DESCRIPTION
**Pull request by threepointone**
via #15319

(redid #15399, removed the changes that didn't even belong in it.)

So tl;dr this PR -

- uses a renderer specific object as the value of ReactCurrentActingRendererSigil.current
- checks this value on 'updateContainer`
- checks this value when setting a state hook's value
- adds a fixture folder for act() (which is run on CI)

This solves 2 specific problems -

- using the wrong act() shouldn't silence the 'missing act' warning
- using the wrong act() logs a warning tha you're, er, using the wrong act()

(Please see  #15399 for the long spiel on the mechanics for this.)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -107,6 +107,7 @@
     "test-prod-build": "yarn test-build-prod",
     "test-build": "cross-env NODE_ENV=development jest --config ./scripts/jest/config.build.js",
     "test-build-prod": "cross-env NODE_ENV=production jest --config ./scripts/jest/config.build.js",
+    "test-dom-fixture": "cd fixtures/dom && yarn && yarn prestart && yarn test",
```
Comment (acdlite) [inline package.json]: Create a CI job for this please. Most people will not remember to run it.

File: packages/react-dom/src/test-utils/ReactTestUtilsAct.js [hunk 2]
```diff
@@ -85,17 +85,19 @@ let actingUpdatesScopeDepth = 0;
...
   if (__DEV__) {
-    ReactShouldWarnActingUpdates.current = true;
+    previousActingUpdatesSigil = ReactActingRendererSigil.current;
+    ReactActingRendererSigil.current = flushPassiveEffects;
   }
 
   function onDone() {
     actingUpdatesScopeDepth--;
     if (__DEV__) {
-      if (actingUpdatesScopeDepth === 0) {
```
Comment (acdlite) [inline packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: I don't get how you were able to remove this depth check. Can you explain?

File: packages/react-dom/src/test-utils/ReactTestUtilsAct.js [hunk 3]
```diff
@@ -85,17 +85,19 @@ let actingUpdatesScopeDepth = 0;
...
 function act(callback: () => Thenable) {
   let previousActingUpdatesScopeDepth = actingUpdatesScopeDepth;
+  let previousActingUpdatesSigil;
   actingUpdatesScopeDepth++;
+  // we use the function flushPassiveEffects directly as the sigil,
+  // since it's unique to a renderer
   if (__DEV__) {
-    ReactShouldWarnActingUpdates.current = true;
+    previousActingUpdatesSigil = ReactActingRendererSigil.current;
+    ReactActingRendererSigil.current = flushPassiveEffects;
```
Comment (acdlite) [inline packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: Could you use a separate export instead? I find this unnecessarily confusing for no benefit, given that it's dev-only.

File: packages/react-dom/src/test-utils/ReactTestUtilsAct.js (hunk 2 above)
Comment (threepointone) [reply packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: ReactActingRendererSigil.current starts with a `.current` value of `null`. Inside `warnIfNotCurrentlyActingUpdatesInDev`, we check whether `.current === flushPassiveEffects)`
As the act() scope opens, it sets the value to flushPassiveEffects which isn't null, *until* the last scope, when previousActingUpdatesSigil gets set as null, triggering the warning.

We could have had the same pattern when the sigil was just a boolean, by saving `.current` as a `previousActingSigilBoolean` or something.

Comment (acdlite) [reply packages/react-dom/src/test-utils/ReactTestUtilsAct.js]: Makes sense

File: package.json (hunk 1 above)
Comment (threepointone) [reply package.json]: I did! It's in https://github.com/facebook/react/pull/15756/files#diff-1d37e48f9ceff6d8030570cd36286a61R167

Comment (acdlite) [reply package.json]: Oh you're right. Missed it!

**Comment by threepointone** [comment]
updated the PR to use an explicit export (from `ReactFiberWorkLoop`) as the value of the sigil. I'll land this later tonight.

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 4]
```diff
@@ -2276,11 +2276,46 @@ function warnAboutInvalidUpdatesOnClassComponentsInDEV(fiber) {
...
+// as the value of ReactCurrentActingRendererSigil.current
+// This identity lets us identify (ha!) when the wrong renderer's act()
+// wraps anothers' updates/effects
+export const ReactActingRendererSigil = {};
+
+export function warnIfNotScopedWithMatchingAct(fiber: Fiber): void {
+  if (__DEV__) {
+    if (
+      ReactCurrentActingRendererSigil.current !== null &&
+      // use the function flushPassiveEffects directly as the sigil
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: Outdated comment

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 5]
```diff
@@ -2276,11 +2276,46 @@ function warnAboutInvalidUpdatesOnClassComponentsInDEV(fiber) {
...
+      // so this comparison is expected here
+      ReactCurrentActingRendererSigil.current !== ReactActingRendererSigil
+    ) {
+      // it looks like we're using the wrong matching act(), so log a warning
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: This comment kinda repeats the warning?

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 6]
```diff
@@ -2276,11 +2276,46 @@ function warnAboutInvalidUpdatesOnClassComponentsInDEV(fiber) {
...
+      warningWithoutStack(
+        false,
+        "It looks like you're using the wrong act() around your test interactions.\n" +
+          'Be sure to use the matching version of act() corresponding to your renderer:\n\n' +
+          '// for react-dom:\n' +
+          "import {act} from 'react-test-utils';\n" +
```
Comment (threepointone) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: there's no such package either 🤦‍♂️

**Pull request merged**